lib/charms/loki_k8s/v1/loki_push_api.py

```diff
@@ -2313,6 +2396,156 @@ def _containers(self) -> Dict[str, Container]:
         return {cont: self._charm.unit.get_container(cont) for cont in self._logs_scheme.keys()}
 
 
+class LogForwarder(ConsumerBase):
+    """Forward the standard outputs of all workloads operated by a charm to one or multiple Loki endpoints."""
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        *,
+        relation_name: str = DEFAULT_RELATION_NAME,
+        loki_endpoints_getter: Optional[Callable[[Relation], List[str]]] = None,
+        alert_rules_path: str = DEFAULT_ALERT_RULES_RELATIVE_PATH,
+        recursive: bool = True,
+        skip_alert_topology_labeling: bool = False,
+    ):
+        super().__init__(
+            charm, relation_name, alert_rules_path, recursive, skip_alert_topology_labeling
+        )
+        self._charm = charm
+        self._relation_name = relation_name
+        self._topology = JujuTopology.from_charm(charm)
+        self._endpoints_getter: Callable = (
+            self._extract_urls if loki_endpoints_getter is None else loki_endpoints_getter
+        )
+
+        on_relation = self._charm.on[self._relation_name]
+        self.framework.observe(on_relation.relation_joined, self._on_logging_relation_joined)
+        self.framework.observe(on_relation.relation_changed, self._on_logging_relation_changed)
+        self.framework.observe(on_relation.relation_departed, self._on_logging_relation_departed)
+        self.framework.observe(on_relation.relation_broken, self._on_logging_relation_broken)
+
+    def _on_logging_relation_joined(self, _):
+        self._enable_logging()
+
+    def _on_logging_relation_changed(self, _):
+        self._enable_logging()
+
+    def _on_logging_relation_departed(self, event: RelationDepartedEvent):
+        self._disable_logging(unit=event.unit)
+
+    def _on_logging_relation_broken(self, event: RelationBrokenEvent):
+        for unit in event.relation.units:
+            if unit.app == self._charm.app:
+                continue
```

Comment:
This should be removed.

Reply:
Actually, this is done because `RelationBrokenEvent` doesn't indicate a single unit, so we need to iterate over all the units in that relation.